src/workarounds/forwarddiff_rules.jl

```diff
@@ -191,54 +191,69 @@ function construct_value(basis::PlaneWaveBasis{T}) where {T <: ForwardDiff.Dual}
                    comm_kpts=basis.comm_kpts)
 end
 
+
 function self_consistent_field(basis_dual::PlaneWaveBasis{T};
                                response=ResponseOptions(),
                                kwargs...) where T <: ForwardDiff.Dual
     # Note: No guarantees on this interface yet.
 
     # Primal pass
-    basis  = construct_value(basis_dual)
-    scfres = self_consistent_field(basis; kwargs...)
-
-    ## promote occupied bands to dual numbers
-    occupation_dual = [T.(occₖ) for occₖ in scfres.occupation]
-    ψ_dual = [Complex.(T.(real(ψₖ)), T.(imag(ψₖ))) for ψₖ in scfres.ψ]
-    ρ_dual = DFTK.compute_density(basis_dual, ψ_dual, occupation_dual)
-    εF_dual = T(scfres.εF)  # Only needed for entropy term
-    eigenvalues_dual = [T.(εₖ) for εₖ in scfres.eigenvalues]
-    energies_dual, ham_dual = energy_hamiltonian(basis_dual, ψ_dual, occupation_dual;
-                                                 ρ=ρ_dual, eigenvalues=eigenvalues_dual,
-                                                 εF=εF_dual)
-
-    response.verbose && println("Solving response problem")
+    basis_primal  = construct_value(basis_dual)
+    scfres = self_consistent_field(basis_primal; kwargs...)
+
+    ## Compute external perturbing potential δV_{ext} (contained in δH)
+    ## and form the matrix-vector product with the bands
+    δH, δHψ = let
+        occupation_dual = [T.(occk) for occk in scfres.occupation]
+        ψ_dual = [Complex.(T.(real(ψk)), T.(imag(ψk))) for ψk in scfres.ψ]
+        ρ_dual = DFTK.compute_density(basis_dual, ψ_dual, occupation_dual)
+        εF_dual = T(scfres.εF)  # Only needed for entropy term
+        eigenvalues_dual = [T.(εk) for εk in scfres.eigenvalues]
+        _, δH = energy_hamiltonian(basis_dual, ψ_dual, occupation_dual;
+                                   ρ=ρ_dual, eigenvalues=eigenvalues_dual, εF=εF_dual)
+        δH, δH * ψ_dual
+    end
 
     ## Implicit differentiation
-    hamψ_dual = ham_dual * ψ_dual
+    response.verbose && println("Solving response problem")
     δresults = ntuple(ForwardDiff.npartials(T)) do α
-        δHψ_α = [ForwardDiff.partials.(δHψk, α) for δHψk in hamψ_dual]
+        δHψ_α = [ForwardDiff.partials.(δHψk, α) for δHψk in δHψ]
+
+        δψ, resp = solve_ΩplusK_split(scfres, -δHψ_α; tol=scfres.norm_Δρ, response.verbose)
+        δρ = compute_δρ(basis_primal, scfres.ψ, δψ, scfres.occupation)
+
+        # TODO Stuff this back into δH to build a δH, by amending the stored total local
+        #      potential. This forms a δH, which has the total variation, which could be
+        #      returned.
+        δVresp   = apply_kernel(basis_primal, δρ; scfres.ρ)  # response potential ε^{-1} V_{ext}
```

Comment:
Hm, I'd rather have this be in solve_omegaplusk_split, because that logic belongs there

Reply:
also this is eps^-1 dV - dV, right?